<a href="https://colab.research.google.com/github/Anavillag/Business-Analytics/blob/main/2_1_Modelo%C3%81rbol.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Modelo Árbol de Decisión**
0. Se cargan las librerias de trabajo

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#Se trae el modelo
from sklearn.tree import DecisionTreeClassifier #Prestamo Interbibliotecario

#Para llamar los archivos del Drive automaticamente - Sin tener que subirlos
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Se cargan los datos desde el drive

In [ ]:
nxl='/content/drive/MyDrive/Analítica de Negocios Ana María Villa/0. SolicitantesCrédito(USD) clase agosto 13.xlsx'
XDB=pd.read_excel(nxl,sheet_name=1)
XDB=XDB.dropna() #Para evitar celdas vacias (no se puede olvidar)
XDB.head()


#Seleccionamos las variables de interés
XD=XDB[['Edad','Ingresos','Egresos','Monto (EAD)']]
XD.head()
XD=np.array(XD)
yd=XDB[['PreApr']];yd=np.array(yd) #Variable de Salida - Convierto a la tabla


2. Se procede conla construcción del modelo de Árbol

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)
ydp=mar.predict(XD)

#Evaluamos el modelo
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]
print('VN=',VN,'FP=',FP,'FN=',FN,'VP=',VP)

Ex=(VP+VN)/(VP+VN+FP+FN)
Ex  #Exactitud

Se=VP/(VP+FN)
Se  #Sensibilidad

Es=VN/(VN+FP)
Es  #Especificidad

print('Exactitud:',Ex)
print('Sensibilidad:',Se)
print('Especificidad:',Es)

VN= 2301 FP= 658 FN= 644 VP= 2239
Exactitud: 0.7771311194796303
Sensibilidad: 0.7766215747485259
Especificidad: 0.7776275768840825


3. Se procede con la visualización del Árbol

In [ ]:
from sklearn.tree import export_graphviz  #Estructura del Árbol
from pydotplus import graph_from_dot_data  #Que datos le voy a colocar al árbol

vs=['Edad','Ingresos','Egresos','Monto (EAD)'] #Estos nombres aparecen en el árbol

dot_data=export_graphviz(mar,feature_names=vs) #Asegurar nombres del gráfico
graph=graph_from_dot_data(dot_data)  #Se hace el gráfico del Árbol
graph.write_png('arbol.png') #Se crea la imagen



True

4. Le calculamos los valores de preaprobación a los solicitantes nuevos

In [ ]:
XDB2=pd.read_excel(nxl,sheet_name=2)
XDB2=XDB2[['Edad','Ingresos','Egresos','Monto (EAD)']]
XDB2=XDB2.dropna() #Para evitar celdas vacias (no se puede olvidar)
XDB2.head()

XD2=np.array(XDB2)
ydp2=mar.predict(XD2)
XDB2['PreApr']=ydp2 #Le pego la columna a la base de datos
XDB2.head()

XDB2.to_excel('SolicitantesPreApr.xlsx')

np.sum(ydp2)


1260

**Análisis**

En la base datos se puede observar un total de 5842 solicitantes de credito divididos en 2959 solicitantes preaprobados (51%) y un total de 2883 solicitantes prenegados (49%). De acuerdo conm lo anterior, el modelo del árbol arrojo un inidce de exactitud cercano al 77%, asi como indices de Sensibilidad cercanos al 77% igualmente y que muestran el comportamiento del modelo frente a la clasifficación de solicitantes PreAprobados. Este mismo porcentaje e promeido se puede observar frente al indice de especificidad el cual mide el comportamiento del modelo en la clasificación de solicitantes PreNegados. Es importante mencionar que un modelo que exhiba valores de exactitud (accuracy) por encima del 75%, es ocnsiderado un modelo aceptable para el prónostico de los datos. Para la base de datos de solicitantes nuevos (2536 solicitantes), se puede observar que el modelo clasififco un total de 1260 solicitantes (%) en la categoria de preaprobados y un total de 1276(%) en la categoria de negados. Lo anterior muestra la predominancia del modelo por la clasificación de la categoria de prenegados.